engine/kv_engine_cleaner.cpp

```diff
@@ -235,366 +237,431 @@ void KVEngine::cleanNoHashIndexedSkiplist(
-          }
-        }
-      }
+    prugeAndFreeAllType(pending_clean_records);
 
-      {  // Destroy skiplist
-        while (!outdated_skip_lists.empty()) {
-          auto& ts_skiplist = outdated_skip_lists.front();
-          if (ts_skiplist.first < version_controller_.LocalOldestSnapshotTS()) {
-            ts_skiplist.second->DestroyAll();
-            removeSkiplist(ts_skiplist.second->ID());
-            outdated_skip_lists.pop_front();
-          } else {
-            break;
-          }
-        }
-      }
+  }  // Finsh iterating hash table
 
-      {  // Destroy list
-        while (!outdated_lists.empty()) {
-          auto& ts_list = outdated_lists.front();
-          if (ts_list.first < version_controller_.LocalOldestSnapshotTS()) {
-            listDestroy(ts_list.second.release());
-            outdated_lists.pop_front();
-          } else {
-            break;
-          }
-        }
-      }
+  // Push the remaining need purged records to global pool.
+  auto new_ts = version_controller_.GetCurrentTimestamp();
+  if (!purge_string_records.empty()) {
+    pending_clean_records.pending_purge_strings.emplace_back(
+        std::move(purge_string_records), new_ts);
+    purge_string_records = std::vector<StringRecord*>();
+  }
 
-      {  // Destroy hash
-        while (!outdated_hash_lists.empty()) {
-          auto& ts_hlist = outdated_hash_lists.front();
-          if (ts_hlist.first < version_controller_.LocalOldestSnapshotTS()) {
-            hashListDestroy(ts_hlist.second.release());
-            outdated_hash_lists.pop_front();
-          } else {
-            break;
-          }
-        }
-      }
-    }  // Finsh iterating hash table
+  if (!purge_dl_records.empty()) {
+    pending_clean_records.pending_purge_dls.emplace_back(
+        std::move(purge_dl_records), new_ts);
+    purge_dl_records = std::vector<DLRecord*>();
+  }
 
-    // Push the remaining need purged records to global pool.
-    auto new_ts = version_controller_.GetCurrentTimestamp();
-    if (!purge_string_records.empty()) {
-      pending_purge_strings.emplace_back(std::move(purge_string_records),
-                                         new_ts);
-      purge_string_records = std::vector<StringRecord*>();
+  return total_num == 0 ? 0.0f : need_purge_num / (double)total_num;
+}
+
+void KVEngine::TestCleanOutDated(size_t start_slot_idx, size_t end_slot_idx) {
+  PendingCleanRecords pending_clean_records;
+  while (!bg_work_signals_.terminating) {
+    auto cur_slot_idx = start_slot_idx;
+    while (cur_slot_idx < end_slot_idx && !bg_work_signals_.terminating) {
+      cleanSlotBlockOutDated(pending_clean_records, cur_slot_idx, 1024);
+      cur_slot_idx += 1024;
     }
+  }
+}
 
-    if (!purge_dl_records.empty()) {
-      pending_purge_dls.emplace_back(std::move(purge_dl_records), new_ts);
-      purge_dl_records = std::vector<DLRecord*>();
+// Space Reclaimer
+void SpaceReclaimer::addNewWorker(size_t thread_id) {
```

Comment:
don't call this addNewWorker, this is worker logic itself

Reply:
change to `doCleanWorker`